provider/vsphere/internal/vsphereclient/createvm.go

```diff
@@ -286,72 +318,181 @@ func uploadImage(
+	}); err != nil {
+		return errors.Trace(err)
 	}
-	return nil
+	return site.Wait()
 }
 
-type progressUpdater struct {
-	tomb   tomb.Tomb
-	ch     chan progress.Report
-	update func(string)
-	action string
+type statusUpdater struct {
+	tomb     tomb.Tomb
+	clock    clock.Clock
+	logger   loggo.Logger
+	ch       chan progress.Report
+	update   func(string)
+	action   string
+	interval time.Duration
 }
 
 // Sink is part of the progress.Sinker interface.
-func (u *progressUpdater) Sink() chan<- progress.Report {
+func (u *statusUpdater) Sink() chan<- progress.Report {
 	return u.ch
 }
 
-func (u *progressUpdater) loop() {
-	defer u.tomb.Done()
-	var last float32
-	const threshold = 10 // update status every X%
+func (u *statusUpdater) loop() error {
+	timer := u.clock.NewTimer(u.interval)
+	defer timer.Stop()
+	var timerChan <-chan time.Time
+
+	var message string
 	for {
 		select {
 		case <-u.tomb.Dying():
-			u.tomb.Kill(tomb.ErrDying)
-			return
+			return tomb.ErrDying
+		case <-timerChan:
+			u.update(message)
+			timer.Reset(u.interval)
+			timerChan = nil
 		case report, ok := <-u.ch:
 			if !ok {
-				return
+				return nil
 			}
-			var message string
 			if err := report.Error(); err != nil {
 				message = fmt.Sprintf("%s: %s", u.action, err)
 			} else {
-				pc := report.Percentage()
-				if pc < 100 && (pc-last) < threshold {
-					// Don't update yet, to avoid spamming
-					// status updates.
-					continue
-				}
-				last = pc
-				message = fmt.Sprintf("%s: %.2f%% (%s)", u.action, pc, report.Detail())
+				message = fmt.Sprintf(
+					"%s: %.2f%% (%s)",
+					u.action,
+					report.Percentage(),
+					report.Detail(),
+				)
 			}
-			u.update(message)
+			timerChan = timer.Chan()
 		}
 	}
 }
 
-func (u *progressUpdater) done() {
+// Kill is part of the worker.Worker interface.
+func (u *statusUpdater) Kill() {
 	u.tomb.Kill(nil)
-	u.tomb.Wait()
+}
+
+// Wait is part of the worker.Worker interface.
+func (u *statusUpdater) Wait() error {
+	return u.tomb.Wait()
+}
+
+type leaseUpdaterContext struct {
+	lease *object.HttpNfcLease
+	start int64
+	size  int64
+	total int64
+}
+
+type leaseUpdater struct {
+	tomb   tomb.Tomb
+	clock  clock.Clock
+	logger loggo.Logger
+	ch     chan progress.Report
+	ctx    context.Context
+	leaseUpdaterContext
+}
+
+// Sink is part of the progress.Sinker interface.
+func (u *leaseUpdater) Sink() chan<- progress.Report {
+	return u.ch
+}
+
+func (u *leaseUpdater) loop() error {
+	const interval = 2 * time.Second
+	timer := u.clock.NewTimer(interval)
+	defer timer.Stop()
+
+	var progress int32
+	for {
+		select {
+		case <-u.tomb.Dying():
+			return tomb.ErrDying
+		case report, ok := <-u.ch:
+			if !ok {
+				return nil
+			}
+			progress = u.progress(report.Percentage())
+		case <-timer.Chan():
+			if err := u.lease.HttpNfcLeaseProgress(u.ctx, progress); err != nil {
+				u.logger.Debugf("failed to update lease progress: %v", err)
+				return errors.Annotate(err, "updating lease progress")
```

Comment:
should this be fatal? maybe a transient thing and will work next try? if there really is an error, won't it fail anyway due to other api calls failing like upload image